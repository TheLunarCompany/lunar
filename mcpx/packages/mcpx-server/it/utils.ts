import {
  accessLogFor,
  buildLogger,
  loggableError,
  LunarLogger,
} from "@mcpx/toolkit-core/logging";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import { Transport } from "@modelcontextprotocol/sdk/shared/transport.js";
import { MeterProvider } from "@opentelemetry/sdk-metrics";
import express from "express";
import { createServer, Server } from "http";
import { ConfigService } from "../src/config.js";
import { Config } from "../src/model/config/config.js";
import { TargetServer } from "../src/model/target-servers.js";
import { AuthGuard, noOpAuthGuard } from "../src/server/auth.js";
import { buildSSERouter } from "../src/server/sse.js";
import { buildStreamableHttpRouter } from "../src/server/streamable.js";
import { buildControlPlaneRouter } from "../src/server/control-plane.js";
import { Services } from "../src/services/services.js";
import {
  TESTKIT_SERVER_CALCULATOR,
  TESTKIT_SERVER_ECHO,
} from "../src/testkit/root.js";

const MCPX_PORT = 9000;

// Track all created loggers for cleanup
const allLoggers = new Set<LunarLogger>();

const getTestLogger: () => LunarLogger = () => {
  const logger = buildLogger({ logLevel: "debug", label: "test" });
  allLoggers.add(logger);
  return logger;
};

export const getMcpxLogger: () => LunarLogger = () => {
  const logger = buildLogger({ logLevel: "debug", label: "mcpx" });
  allLoggers.add(logger);
  return logger;
};

// Cleanup function to close all logger instances (for global teardown)
export const closeAllLoggers = (): void => {
  allLoggers.forEach((logger) => {
    try {
      logger.close();
    } catch (_e) {
      // Ignore errors when closing loggers
    }
  });
  allLoggers.clear();
};

const BASE_CONFIG: Config = {
  permissions: {
    default: { _type: "default-allow", block: [] },
    consumers: {},
  },
  toolGroups: [],
  auth: { enabled: false },
  toolExtensions: { services: {} },
};
export function buildConfig(props: Partial<Config> = {}): ConfigService {
  const { auth, permissions, toolGroups, toolExtensions } = {
    ...BASE_CONFIG,
    ...props,
  };
  return new ConfigService(
    { permissions, toolGroups, auth, toolExtensions },
    getMcpxLogger(),
  );
}

export const stdioTargetServers: TargetServer[] = [
  {
    type: "stdio",
    name: "echo-service",
    command: "node",
    args: [TESTKIT_SERVER_ECHO],
    env: {},
  },
  {
    type: "stdio",
    name: "calculator-service",
    command: "node",
    args: [TESTKIT_SERVER_CALCULATOR],
    env: {},
  },
];
export const oauthTargetServer: TargetServer = {
  type: "streamable-http",
  name: "oauth-mock-server",
  url: "http://localhost:9001/mcp",
};
export const allTargetServers: TargetServer[] = [
  ...stdioTargetServers,
  oauthTargetServer,
];
type TransportType = "SSE" | "StreamableHTTP";

export const transportTypes: TransportType[] = [
  "SSE",
  "StreamableHTTP",
] as const;

export class TestHarness {
  public clientConnectError?: Error | undefined;
  private loggers: LunarLogger[] = [];

  constructor(
    public client: Client,
    public server: Server,
    public services: Services,
    public testLogger: LunarLogger,
    private clientConnectExtraHeaders: Record<string, string> = {},
    private targetServers: TargetServer[] = stdioTargetServers,
  ) {
    // Track loggers for this harness instance
    this.loggers.push(testLogger);
  }

  addLogger(logger: LunarLogger): void {
    this.loggers.push(logger);
  }

  async initialize(transportType: TransportType): Promise<void> {
    // Setup MCPX
    await this.services.initialize();
    await Promise.all(
      this.targetServers.map((target) =>
        this.services.targetClients.addClient(target),
      ),
    );
    await this.server.listen(MCPX_PORT, () => {
      this.testLogger.info(`Test MCPX server listening on port ${MCPX_PORT}`);
    });

    const transport = this.buildTransport(transportType);

    await this.client
      .connect(transport)
      .then(() => this.testLogger.info("End-client connected"))
      .catch((e) => {
        const error = loggableError(e);
        this.testLogger.error("Failed to connect end-client", error);
        this.clientConnectError = e;
      });
  }

  async shutdown(): Promise<void> {
    await this.client.transport?.close();
    await this.client.close();
    await this.services.shutdown();
    await new Promise<void>((resolve) => {
      this.server.close(() => {
        this.testLogger.info("Test MCPX server closed");
        // Close loggers AFTER server has fully closed and all connection events have fired
        this.loggers.forEach((logger) => {
          try {
            logger.close();
          } catch (_e) {
            // Ignore errors when closing loggers
          }
        });
        this.loggers = [];
        resolve();
      });
    });
  }

  private buildTransport(transportType: TransportType): Transport {
    switch (transportType) {
      case "SSE":
        return new SSEClientTransport(
          new URL(`http://localhost:${MCPX_PORT}/sse`),
          {
            eventSourceInit: {
              fetch: (url, init) => {
                const headers = new Headers({
                  ...init?.headers,
                  ...this.clientConnectExtraHeaders,
                });
                return fetch(url, { ...init, headers });
              },
            },
          },
        );
      case "StreamableHTTP":
        return new StreamableHTTPClientTransport(
          new URL(`http://localhost:${MCPX_PORT}/mcp`),
          {
            requestInit: {
              headers: this.clientConnectExtraHeaders,
            },
          },
        );
      default:
        throw new Error(`Unknown transport type: ${transportType}`);
    }
  }
}

interface TestHarnessProps {
  config?: ConfigService;
  authGuard?: AuthGuard;
  mcpxLogger?: LunarLogger;
  clientConnectExtraHeaders?: Record<string, string>;
  targetServers?: TargetServer[];
}
function defaultTestHarnessProps(): Required<TestHarnessProps> {
  return {
    config: buildConfig(),
    authGuard: noOpAuthGuard,
    mcpxLogger: getMcpxLogger(),
    clientConnectExtraHeaders: {},
    targetServers: stdioTargetServers,
  };
}
export function getTestHarness(props: TestHarnessProps = {}): TestHarness {
  // the effective values are the defaults, unless overridden by props
  const {
    config,
    authGuard,
    mcpxLogger,
    clientConnectExtraHeaders,
    targetServers,
  } = {
    ...defaultTestHarnessProps(),
    ...props,
  };
  const testLogger = getTestLogger();
  const meterProvider = new MeterProvider();
  const services = new Services(config, meterProvider, testLogger);

  const client = new Client({ name: "end-client", version: "1.0.0" });

  // Wire router on server
  const sseRouter = buildSSERouter(authGuard, services, mcpxLogger);
  const streamableRouter = buildStreamableHttpRouter(
    authGuard,
    services,
    mcpxLogger,
  );
  const controlPlaneRouter = buildControlPlaneRouter(
    authGuard,
    services,
    mcpxLogger,
  );
  const app = express();
  const httpServer = createServer(app);
  app.use(express.json());
  app.use(accessLogFor(mcpxLogger));
  app.use(sseRouter);
  app.use(streamableRouter);
  app.use(controlPlaneRouter);

  const harness = new TestHarness(
    client,
    httpServer,
    services,
    testLogger,
    clientConnectExtraHeaders,
    targetServers,
  );

  // Track the mcpxLogger in this harness instance
  harness.addLogger(mcpxLogger);

  return harness;
}
