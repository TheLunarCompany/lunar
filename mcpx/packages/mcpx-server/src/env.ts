import "dotenv/config";
import { z } from "zod/v4";

const envSchema = z.object({
  API_KEY: z.string().optional(),
  PORT: z.coerce.number().default(9000),
  WEBSERVER_PORT: z.coerce.number().default(9001),
  ENABLE_WEBSERVER: z.coerce.boolean().default(false),
});

export type Env = z.infer<typeof envSchema>;

export const env = envSchema.parse(process.env);
