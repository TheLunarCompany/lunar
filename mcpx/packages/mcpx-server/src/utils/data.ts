// A function to remove `null`s and/or `undefined`s from an array
export function compact<X>(xs: Array<X | null | undefined>): X[] {
  return xs.flatMap((x) => (x ? [x] : []));
}

export function stringifyEq<T>(a: T, b: T): boolean {
  return JSON.stringify(a) === JSON.stringify(b);
}
