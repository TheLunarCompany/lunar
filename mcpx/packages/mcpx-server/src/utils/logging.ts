import { makeError } from "./errors.js";

export interface LoggableError {
  name: string;
  message: string;
  stack?: string;
}

export function loggableError(e: unknown): LoggableError {
  const error = makeError(e);
  return {
    name: error.name,
    message: error.message,
    stack: error.stack,
  };
}
