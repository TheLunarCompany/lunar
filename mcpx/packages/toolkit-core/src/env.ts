import "dotenv/config";
import { z } from "zod/v4";

const envSchema = z.object({
  LOKI_HOST: z.string().default("log-collector-dev.lunar.dev"),
  VERSION: z.string().default("0.0.0"),
  INSTANCE_ID: z.string().default("unknown"),
  LUNAR_TELEMETRY: z.stringbool().default(true),
  LUNAR_API_KEY: z.string().default(""),
});

export type Env = z.infer<typeof envSchema>;

export const env = envSchema.parse(process.env);
