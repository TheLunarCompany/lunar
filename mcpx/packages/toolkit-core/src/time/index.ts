export * from "./clock.js";
export * from "./measure.js";
export * from "./polling.js";
