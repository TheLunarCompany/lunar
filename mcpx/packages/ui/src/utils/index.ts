export function createPageUrl(pageName: string) {
  return "/" + pageName.toLowerCase().replace(/ /g, "-");
}
