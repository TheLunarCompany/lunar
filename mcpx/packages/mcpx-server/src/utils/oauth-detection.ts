import { loggableError } from "@mcpx/toolkit-core/logging";
import { Logger } from "winston";
import z from "zod/v4";
import { RemoteTargetServer } from "../model/target-servers.js";

// These schemas are based on the OAuth 2.0 spec and conventions
const protectedResourceSchema = z.object({
  authorization_servers: z.array(z.unknown()).min(1),
});

const authServerSchema = z.object({
  authorization_endpoint: z.string().optional(),
  token_endpoint: z.string().optional(),
});

// Type for the fetch function to enable dependency injection
export type FetchFn = typeof fetch;

// Check if a remote server supports OAuth by looking for well-known endpoints
// Note that this implementation is based on the assumption that MCP servers will
// implement the OAuth spec correctly, or at least follow some of the conventions.
// This is a legitimate assumption since the MCP TS SDK uses this approach as well,
// to the best of our knowledge.
export async function detectOAuthSupport(
  targetServer: RemoteTargetServer,
  discoveryTimeoutMillis: number,
  logger: Logger,
  fetchFn: FetchFn = fetch,
): Promise<boolean> {
  try {
    const baseUrl = new URL(targetServer.url).origin;

    // Check both endpoints in parallel
    const [protectedResourceResult, authServerResult] =
      await Promise.allSettled([
        // Check protected resource metadata (correct per MCP spec, e.g. Notion)
        fetchFn(`${baseUrl}/.well-known/oauth-protected-resource`, {
          method: "GET",
          signal: AbortSignal.timeout(discoveryTimeoutMillis),
        }).then(async (response) => {
          if (!response.ok) return false;
          const json = await response.json();
          const parsed = protectedResourceSchema.safeParse(json);
          return parsed.success;
        }),

        // Check authorization server metadata (sometimes used by MCP servers, e.g. Linear)
        fetchFn(`${baseUrl}/.well-known/oauth-authorization-server`, {
          method: "GET",
          signal: AbortSignal.timeout(discoveryTimeoutMillis),
        }).then(async (response) => {
          if (!response.ok) return false;
          const json = await response.json();
          const parsed = authServerSchema.safeParse(json);
          return (
            parsed.success &&
            Boolean(
              parsed.data.authorization_endpoint || parsed.data.token_endpoint,
            )
          );
        }),
      ]);

    // If either result is `fulfilled` (i.e. successful HTTP call, including schema validation),
    // and valid, OAuth is assumed to be supported
    const protectedResourceSupported =
      protectedResourceResult.status === "fulfilled" &&
      protectedResourceResult.value;
    const authServerSupported =
      authServerResult.status === "fulfilled" && authServerResult.value;

    const oauthSupported = protectedResourceSupported || authServerSupported;

    logger.debug(
      `OAuth support ${oauthSupported ? "was" : "was not"} detected`,
      {
        name: targetServer.name,
        url: targetServer.url,
        basedOn: {
          protectedResource: protectedResourceSupported,
          authServer: authServerSupported,
        },
      },
    );
    return oauthSupported;
  } catch (error) {
    logger.debug("Error checking OAuth support", {
      name: targetServer.name,
      error: loggableError(error),
    });
    return false;
  }
}
