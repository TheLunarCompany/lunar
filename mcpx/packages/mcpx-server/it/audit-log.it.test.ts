import { afterEach, beforeEach, describe, it } from "@jest/globals";
import { withAsyncPolling } from "@mcpx/toolkit-core/time";
import fs from "fs";
import path from "path";
import { env, resetEnv } from "../src/env.js";
import { getTestHarness, transportTypes } from "./utils.js";

describe.each(transportTypes)("Audit Log Service over %s", (transportType) => {
  let testHarness: ReturnType<typeof getTestHarness>;

  const originalEnv = { ...process.env };

  beforeEach(async () => {
    process.env = { ...originalEnv };
    process.env["AUDIT_LOG_FLUSH_INTERVAL_IN_SEC"] = "1";
    resetEnv();
    await deleteAllAuditLogFiles();
    testHarness = getTestHarness();
    await testHarness.initialize(transportType);
  });

  afterEach(async () => {
    process.env = { ...originalEnv };
    resetEnv();
    await deleteAllAuditLogFiles();
    await testHarness.shutdown();
  });

  it("should persist events to file", async () => {
    await testHarness.client.callTool({
      name: "echo-service__echo",
      arguments: { message: "The sound of silence?" },
    });

    // poll for 10 seconds
    const auditFile = await withAsyncPolling({
      maxAttempts: 200,
      sleepTimeMs: 50,
      getValue: async () => await findLatestAuditLogFile(),
      found: (fileContent): fileContent is string => Boolean(fileContent),
    });

    console.log("Audit file content:", auditFile);
    expect(auditFile).toBeDefined();
    // split by lines
    const lines = auditFile.split("\n").filter(Boolean);
    expect(lines.length).toBe(2); // 1 for initial config applied, 1 for echo service call
    const configAppliedEvent = JSON.parse(lines[0]!);
    const toolUsedEvent = JSON.parse(lines[1]!);

    // Both should be written at the same flush - createdAt should be the same
    const configAppliedCreatedAt = Date.parse(configAppliedEvent.createdAt);
    const toolUsedCreatedAt = Date.parse(toolUsedEvent.createdAt);
    expect(configAppliedCreatedAt).toEqual(toolUsedCreatedAt);

    // However timestamp of each event is different as they happen at different times
    const configAppliedTimestamp = Date.parse(configAppliedEvent.timestamp);
    const toolUsedTimestamp = Date.parse(toolUsedEvent.timestamp);
    expect(configAppliedTimestamp).not.toEqual(toolUsedTimestamp);

    // Applied Config Content
    expect(configAppliedEvent.eventType).toBe("config_applied");
    expect(configAppliedEvent.payload.version).toEqual(1);

    // Tool Used Content
    expect(toolUsedEvent.eventType).toBe("tool_used");
    expect(toolUsedEvent.payload).toMatchSnapshot();
  });
});

async function deleteAllAuditLogFiles(): Promise<void> {
  try {
    const files = await fs.promises.readdir(env.AUDIT_LOG_DIR);
    await Promise.all(
      files.map((file) =>
        fs.promises.unlink(path.join(env.AUDIT_LOG_DIR, file)),
      ),
    );
    console.log("All files deleted successfully", { count: files.length });
  } catch (err) {
    console.error("Error deleting files:", err);
  }
}

async function findLatestAuditLogFile(): Promise<string | null> {
  try {
    const files = await fs.promises.readdir(env.AUDIT_LOG_DIR);
    if (files.length === 0) {
      return null;
    }

    const fileStats = await Promise.all(
      files.map(async (file) => {
        const filePath = path.join(env.AUDIT_LOG_DIR, file);
        const stats = await fs.promises.stat(filePath);
        return { file, time: stats.mtime };
      }),
    );

    // Find the latest file
    const latest = fileStats.reduce((a, b) => (a.time > b.time ? a : b));

    // Read and return its content
    const content = await fs.promises.readFile(
      path.join(env.AUDIT_LOG_DIR, latest.file),
      "utf-8",
    );
    console.log("Latest file read successfully:", path.basename(latest.file));
    return content;
  } catch (err) {
    console.error("Error finding latest file:", err);
    return null;
  }
}
