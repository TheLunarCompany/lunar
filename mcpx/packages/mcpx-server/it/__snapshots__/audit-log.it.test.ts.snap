// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit Log Service over SSE should persist events to file 1`] = `
{
  "args": {
    "message": "The sound of silence?",
  },
  "targetServerName": "echo-service",
  "toolName": "echo",
}
`;

exports[`Audit Log Service over StreamableHTTP should persist events to file 1`] = `
{
  "args": {
    "message": "The sound of silence?",
  },
  "targetServerName": "echo-service",
  "toolName": "echo",
}
`;
